sky/shell/android/update_service_android.cc

```diff
@@ -0,0 +1,126 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "base/android/jni_string.h"
+#include "base/bind.h"
+#include "base/files/file_util.h"
+#include "base/logging.h"
+#include "base/threading/worker_pool.h"
+#include "jni/UpdateService_jni.h"
+#include "mojo/data_pipe_utils/data_pipe_utils.h"
+#include "mojo/public/cpp/application/connect.h"
+#include "mojo/public/interfaces/application/service_provider.mojom.h"
+#include "mojo/services/network/public/interfaces/network_service.mojom.h"
+#include "sky/shell/shell.h"
+
+namespace sky {
+namespace shell {
+
+namespace {
+
+// TODO(mpcomplete): make this a utility method?
+static mojo::URLLoaderPtr FetchURL(
+    mojo::NetworkService* network_service,
+    const std::string& url,
+    base::Callback<void(mojo::URLResponsePtr)> callback) {
+  mojo::URLLoaderPtr loader;
+  network_service->CreateURLLoader(GetProxy(&loader));
+
+  mojo::URLRequestPtr request = mojo::URLRequest::New();
+  request->url = url;
+  request->auto_follow_redirects = true;
+  loader->Start(request.Pass(), callback);
+
+  return loader.Pass();
+}
+
+class UpdateTask {
+ public:
+  UpdateTask(JNIEnv* env, jobject update_service, std::string data_dir)
+      : final_path_(base::FilePath(data_dir).AppendASCII("app.skyx")) {
+    update_service_.Reset(env, update_service);
+
+    mojo::ServiceProviderPtr service_provider =
+        CreateServiceProvider(Shell::Shared().service_provider_context());
+    mojo::ConnectToService(service_provider.get(), &network_service_);
+  }
+
+  void DownloadSkyx(const std::string& url) {
```

Comment:
nit: s/DownloadSkyx/DownloadAppBundle/

skyx is just the file extension.  We've been calling the thing a "bundle" and in particular an "app bundle" if its a bundle that contains an app.